### PR TITLE
Feature/add drag overlay

### PR DESCRIPTION
**Pull request by AdityaKhatri**
- Addresses https://github.com/the-deep/client/issues/1685

## Changes

* Create SortableList component

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments

## This PR contains valid:

- [x] permission checks
- [x] translations

> NOTE: I've commented out some code which is necessary at a later point, plz ignore them

### CODE REVIEW

File: src/utils/safeCommon.ts [hunk 1]
```diff
@@ -1,8 +1,12 @@
+import { memo } from 'react';
+
 import {
     isDefined,
     isTruthyString,
 } from '@togglecorp/fujs';
 
+export const genericMemo: (<T>(c: T) => T) = memo;
```
Comment (subinasr) [inline src/utils/safeCommon.ts:9]: I don't understand this. Explain please?

**Review by subinasr — CHANGES_REQUESTED** [review]

File: src/components/ui/SortableList/index.tsx [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    DragStartEvent,
+    DraggableSyntheticListeners,
+} from '@dnd-kit/core';
+import { CSS } from '@dnd-kit/utilities';
+/*
+import {
+    restrictToHorizontalAxis,
+    restrictToVerticalAxis,
+} from '@dnd-kit/modifiers';
+*/
```
Comment (samshara) [inline src/components/ui/SortableList/index.tsx]: remove this, if important add a note to explain why the code is commented out.

File: src/components/ui/SortableList/index.tsx [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+
+    return (
+        <DndContext
+            sensors={sensors}
+            collisionDetection={closestCenter}
+            onDragStart={handleDragStart}
+            onDragEnd={handleDragEnd}
+            // modifiers={[
+            //     direction === 'horizontal' ? restrictToHorizontalAxis : restrictToVerticalAxis,
+            // ]}
```
Comment (samshara) [inline src/components/ui/SortableList/index.tsx]: remove this

File: src/components/ui/SortableList/index.tsx [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        activeId,
+        Renderer,
+        keySelector,
+        rendererParams,
+        data,
+        showDragOverlay,
+    ]);
+
+    const modifiedRendererParams = useCallback((
+        key: K,
```
Comment (samshara) [inline src/components/ui/SortableList/index.tsx]:
```suggestion
        _: K,
```

File: src/utils/safeCommon.ts (hunk 1 above)
Comment (AdityaKhatri) [reply src/utils/safeCommon.ts:9]: TS issues will occur if the memo is directly used in components that use [typescript generics](https://www.typescriptlang.org/docs/handbook/2/generics.html). This is done to make it type safe.
@tnagorra could you add more details?

**Review by samshara — APPROVED** [review]

File: src/utils/safeCommon.ts (hunk 1 above)
Comment (tnagorra) [reply src/utils/safeCommon.ts:9]: I could if I knew better!

File: src/components/ui/SortableList/index.tsx [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+interface SortableItemProps<D, P, K extends OptionKey> {
+    keySelector: (data: D) => K;
+    datum: D;
+    renderer: (props: P & {
+        listeners?: Listeners;
+        attributes?: Attributes;
+        setNodeRef?: NodeRef;
+        style?: React.CSSProperties;
+    }) => JSX.Element;
+    params: P;
```
Comment (tnagorra) [inline src/components/ui/SortableList/index.tsx]:
```suggestion
    rendererParams: P;
```

File: src/components/ui/SortableList/index.tsx [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+
+    const style: React.CSSProperties = {
+        transform: CSS.Transform.toString({
+            x: transform?.x ?? 0,
+            y: transform?.y ?? 0,
+            scaleX: 1,
+            scaleY: 1,
+        }),
+        transition: transition ?? undefined,
+    };
```
Comment (tnagorra) [inline src/components/ui/SortableList/index.tsx]: Why not use useMemo?

File: src/components/ui/SortableList/index.tsx [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+            sensors={sensors}
+            collisionDetection={closestCenter}
+            onDragStart={handleDragStart}
+            onDragEnd={handleDragEnd}
+            modifiers={[
+                direction === 'horizontal' ? restrictToHorizontalAxis : restrictToVerticalAxis,
+            ]}
+        >
+            <SortableContext
+                items={items ?? []}
```
Comment (tnagorra) [inline src/components/ui/SortableList/index.tsx]:
```suggestion
                items={items ?? []}
```
Let's not create new elements on render

File: src/components/ui/SortableList/index.tsx [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+                    className={className}
+                    data={data}
+                    keySelector={keySelector}
+                    renderer={MemoizedSortableItem}
+                    rendererParams={modifiedRendererParams}
+                    {...otherProps}
+                    grouped={false}
+                />
+            </SortableContext>
+            {showDragOverlay && createPortal((
```
Comment (tnagorra) [inline src/components/ui/SortableList/index.tsx]: We already have a Portal component somewhere

File: src/newViews/AnalyticalFramework/PrimaryTagging/index.tsx [hunk 9]
```diff
@@ -439,7 +444,7 @@ function PrimaryTagging(props: Props) {
 
     const appliedSections = useMemo(
         () => {
-            const mySections = tempSections ?? sections;
+            const mySections = (tempSections ?? sections).sort(compareOrder);
```
Comment (tnagorra) [inline src/newViews/AnalyticalFramework/PrimaryTagging/index.tsx]: Sort will mutate data

**Review by tnagorra — CHANGES_REQUESTED** [review]

**Review by tnagorra — APPROVED** [review]

**Pull request merged**